infra/Pos/Slotting/Impl/Util.hs

```diff
@@ -15,57 +15,104 @@ import           Pos.Core.Slotting           (flattenEpochIndex, unflattenSlotId
 import           Pos.Core.Timestamp          (addTimeDiffToTimestamp)
 import           Pos.Core.Types              (EpochIndex, SlotId (..), Timestamp (..),
                                               mkLocalSlotIndex)
-import           Pos.Util.Util               (leftToPanic)
+import           Pos.Util.Util               (leftToPanic, maybeThrow)
 
+import           Pos.Slotting.Error          (SlottingError (..))
 import           Pos.Slotting.MemState.Class (MonadSlotsData (..))
-import           Pos.Slotting.Types          (EpochSlottingData (..), SlottingData (..))
+import           Pos.Slotting.Types          (EpochSlottingData (..))
 
 -- | Approximate current slot using outdated slotting data.
-approxSlotUsingOutdated :: MonadSlotsData m => EpochIndex -> Timestamp -> m SlotId
-approxSlotUsingOutdated penult t = do
-    SlottingData {..} <- getSlottingData
-    systemStart <- getSystemStart
-    let epochStart = esdStartDiff sdLast `addTimeDiffToTimestamp` systemStart
+approxSlotUsingOutdated
+    :: (MonadSlotsData m)
+    => Timestamp
+    -> m SlotId
+approxSlotUsingOutdated t = do
+
+    systemStart  <- getSystemStartM
+    currentIndex <- getCurrentEpochIndexM
+    sdNext       <- getNextEpochSlottingDataM
```

Comment:
Good point. Didn't occur to me, but it can happen. I could use a tuple or a structure for returning them both at the same time, and add a comment?
I would like to keep `STM` local in that typeclass, keeps things simple(r).